src/Kokkos_HalfPrecision.hpp

```diff
@@ -0,0 +1,173 @@
+         * results via a KOKKOSKERNELS_HAVE define to KokkosKernels_config.h
+         * 
+         * First we check for cuda half precision support   (HAVE_KOKKOSKERNELS_CUDA_FP16).
+         * Second we check for host half precision support  (HAVE_KOKKOSKERNELS_FP16).
+         * Lastly, we fall back to single precision support.
+         * 
+         * NOTE: If both cuda and host support half precision, the half type will
+         * default to device_fp16_t.
+         */
+        #if defined(KOKKOS_ENABLE_CUDA) && defined(HAVE_KOKKOSKERNELS_CUDA_FP16)
+            using device_fp16_t = __half;
+            using half = device_fp16_t;
+            #if defined(HAVE_KOKKOSKERNELS_FP16)
+                using host_fp16_t = _Float16;
+            #else
+                using host_fp16_t = float;
+            #endif // defined(HAVE_KOKKOSKERNELS_FP16)
+            static KOKKOS_FORCEINLINE_FUNCTION float __cast2float(device_fp16_t x) { return __half2float(x); }
+            static KOKKOS_FORCEINLINE_FUNCTION device_fp16_t __cast2half(float x) { return __float2half(x); }
+            
+        #else // defined(KOKKOS_ENABLE_CUDA) && defined(HAVE_KOKKOSKERNELS_CUDA_FP16)
+            #if defined(HAVE_KOKKOSKERNELS_FP16)
+                using host_fp16_t = _Float16;
+                using half = host_fp16_t;
+            #else // defined(HAVE_KOKKOSKERNELS_FP16)
+                using host_fp16_t = float;
+                using device_fp16_t = host_fp16_t;
+                using half = host_fp16_t;
+            #endif // _Float16
+            static inline float __cast2float(host_fp16_t x) { return (float) x; }
+            static inline host_fp16_t __cast2half(float x) { return (host_fp16_t) x; }
+        #endif
+        ////////////// BEGIN half2float and float2half overloads //////////////
+        /**
+        * Since kokkos does not have support for half precision types yet, we 
+        * must cast to/from float in some kokkos-kernels routines. Except for
+        * the overloads below that actually cast to/from half precision types,
+        * the others should be optimized away by the compiler.
+        */
+        // host_fp16_t
+#if defined(HAVE_KOKKOSKERNELS_FP16)
+        static inline
+        float half2float(host_fp16_t x, float &ret) {
+            ret = __cast2float(x);
+            return ret;
+        }
+        static inline
+        host_fp16_t float2half(float x, host_fp16_t &ret) {
+            ret = __cast2half(x);
+            return ret;
+        }
+#endif
+        // device_fp16_t
+#if defined(KOKKOS_ENABLE_CUDA) && defined(HAVE_KOKKOSKERNELS_CUDA_FP16)
+        static inline __device__
+        float half2float(device_fp16_t x, float &ret) {
+            ret = __cast2float(x);
+            return ret;
+        }
+        static inline __device__
+        device_fp16_t float2half(float x, device_fp16_t &ret) {
+            ret = __cast2half(x);
+            return ret;
+        }
+#endif
+        // float
```

Comment:
The following overloads feel funky. Even if they are temporary workaround you may get into trouble. Consider just having generic functions:
```
half cast_to_half(double); half cast_to_half(float); half cast_to_half(half);
```

Reply:
With trouble I mean later, when you try to get rid of them again.

Reply:
Also generally having a cast function both take a return argument and return it is pretty iffy from a design perspective.

Reply:
if you need the other direction add cast_to_double etc. Note I get that this float2half business is how CUDA names the cast functions but there are not different overloads of those functions. I would REALLY recommend to change this in particular since you make them public. We could also deal with this inside of Kokkos Core and do a proper solution for you.

Reply:
agreed. these overloads can actually be removed since GEMM is the only kernel being used with `half`. the overloads were added [here](https://github.com/e10harvey/kokkos-kernels/blob/6eda84d2b9cd675cc182385d6627297ceb53a961/src/batched/KokkosBatched_FindAmax_Internal.hpp#L62) as preliminary work around for [this](https://github.com/e10harvey/kokkos-kernels/blob/6eda84d2b9cd675cc182385d6627297ceb53a961/src/batched/KokkosBatched_FindAmax_Internal.hpp#L50-L53). i'd be interested in helping with the proper solution within Kokkos Core. I will remove these overloads unless I hear otherwise.